File: scripts/parity/schema/ruby/canonicalize.rb
# frozen_string_literal: true

require "set"
require "bigdecimal"

# Lowers a native dump (raw AR introspection data) into the neutral
# CanonicalSchema format defined in scripts/parity/canonical/schema.schema.json.
#
# Pure function — no I/O, no side effects.

module Canonicalize
  FILTERED_TABLES = %w[schema_migrations ar_internal_metadata].to_set.freeze
  AUTOINDEX_PREFIX = "sqlite_autoindex_"

  # Maps AR abstract type symbols (col.type) to canonical type strings.
  # D4: throw on any type not in this map.
  AR_TO_CANONICAL = {
    string:   "string",
    text:     "text",
    integer:  "integer",
    bigint:   "bigint",
    float:    "float",
    decimal:  "decimal",
    datetime: "datetime",
    date:     "date",
    time:     "time",
    boolean:  "boolean",
    binary:   "binary",
    json:     "json",
  }.freeze

  # native_dump: Hash<table_name, { columns:, indexes:, primary_key_columns: }>
  # where columns is Array<{ name:, ar_type:, null:, default:, limit:, precision:, scale: }>
  # and indexes is Array<{ name:, columns:, unique:, where: }>
  # and primary_key_columns is Array<String> in PK-position order
  def self.call(native_dump)
    tables = native_dump
      .reject { |name, _| FILTERED_TABLES.include?(name) }
      .sort_by { |name, _| name }
      .map { |name, table| canonicalize_table(name, table) }

    { "version" => 1, "tables" => tables }
  end

  def self.canonicalize_table(name, table)
    pk_cols = table[:primary_key_columns]
    primary_key = case pk_cols.length
    when 0 then nil
    when 1 then pk_cols[0]
    else        pk_cols
    end

    columns = table[:columns].map { |col| canonicalize_column(name, col) }

    indexes = table[:indexes]
      .reject { |idx| idx[:name].start_with?(AUTOINDEX_PREFIX) }
      .sort_by { |idx| idx[:name] }
      .map { |idx| canonicalize_index(name, idx) }

    { "name" => name, "primaryKey" => primary_key, "columns" => columns, "indexes" => indexes }
  end

  def self.canonicalize_column(table_name, col)
    ar_type = col[:ar_type]
    canonical_type = AR_TO_CANONICAL[ar_type] or
      raise "canonicalize: unknown AR type #{ar_type.inspect} on #{table_name}.#{col[:name]} — add it to AR_TO_CANONICAL"

    {
      "name"      => col[:name],
      "type"      => canonical_type,
      "null"      => col[:null],
      "default"   => coerce_default(col[:default]),
      "limit"     => col[:limit],
      "precision" => col[:precision],
      "scale"     => col[:scale],
    }
  end

  def self.canonicalize_index(table_name, idx)
    cols = idx[:columns]
    raise "canonicalize: index #{idx[:name].inspect} on #{table_name} has no columns" if cols.empty?

    {
      "name"    => idx[:name],
      "columns" => cols,
      "unique"  => idx[:unique],
      "where"   => idx[:where],
    }
  end

  # Coerce an AR default value (already deserialized by AR's type system)
  # into a canonical scalar: String | Integer | Float | true | false | nil.
  def self.coerce_default(value)
    return nil if value.nil?
    return value if value == true || value == false
    return value.to_i if value.is_a?(Integer)
    # Float and BigDecimal (AR uses BigDecimal for :decimal columns) → JSON number.
    return value.to_f if value.is_a?(Numeric)
    value.to_s
  end
end

File: scripts/parity/schema/node/canonicalize.ts
/**
 * Lowers a native dump (raw adapter introspection data) into the neutral
 * CanonicalSchema format defined in scripts/parity/canonical/schema.schema.json.
 *
 * Pure function — no I/O, no side effects.
 */

import type {
  CanonicalColumn,
  CanonicalIndex,
  CanonicalSchema,
  CanonicalTable,
  CanonicalType,
} from "../../canonical/types.js";

export interface NativeColumn {
  name: string;
  /** Raw SQL type string as returned by PRAGMA table_info (e.g. "INTEGER", "TEXT"). */
  sqlType: string;
  primaryKey: boolean;
  /** true = nullable */
  null: boolean;
  /** Raw default string from PRAGMA table_info dflt_value, or null. */
  default: string | null;
  limit: number | null;
  precision: number | null;
  scale: number | null;
}

export interface NativeIndex {
  name: string;
  columns: string[];
  unique: boolean;
  where?: string | null;
}

export interface NativeTable {
  /** Columns in declaration order (as returned by introspectColumns). */
  columns: NativeColumn[];
  indexes: NativeIndex[];
  /**
   * Primary key column names in PK-position order (from adapter.primaryKey,
   * which sorts by PRAGMA table_info `pk` field). Empty array = no PK.
   */
  primaryKeyColumns: string[];
}

/** Output shape of dump.ts — keyed by table name. */
export type NativeDump = Record<string, NativeTable>;

// SQLite PRAGMA type strings → canonical CanonicalType.
// Keys are lowercased. Only types that appear in SQLite fixtures are
// required here; extend when adding PG/MySQL fixtures.
const SQL_TO_CANONICAL: Record<string, CanonicalType> = {
  integer: "integer",
  int: "integer",
  bigint: "bigint",
  text: "text",
  varchar: "string",
  "character varying": "string",
  real: "float",
  float: "float",
  double: "float",
  "double precision": "float",
  numeric: "decimal",
  decimal: "decimal",
  blob: "binary",
  binary: "binary",
  bytea: "binary",
  boolean: "boolean",
  bool: "boolean",
  datetime: "datetime",
  timestamp: "datetime",
  date: "date",
  time: "time",
  json: "json",
  jsonb: "json",
};

const FILTERED_TABLES = new Set(["schema_migrations", "ar_internal_metadata"]);
const AUTOINDEX_PREFIX = "sqlite_autoindex_";

function toCanonicalType(sqlType: string, table: string, column: string): CanonicalType {
  const base = sqlType
    .toLowerCase()
    .trim()
    .replace(/\s*\([^)]*\)/, "");
  const mapped = SQL_TO_CANONICAL[base];
  if (!mapped) {
    throw new Error(
      `canonicalize: unknown SQL type "${sqlType}" on ${table}.${column} — add it to SQL_TO_CANONICAL`,
    );
  }
  return mapped;
}

/** Coerce a raw PRAGMA dflt_value string to a canonical scalar. */
function coerceDefault(raw: string | null): string | number | boolean | null {
  if (raw === null) return null;
  // Strip surrounding single quotes: 'hello' → "hello"
  if (raw.startsWith("'") && raw.endsWith("'")) {
    return raw.slice(1, -1).replace(/''/g, "'");
  }
  if (raw === "true") return true;
  if (raw === "false") return false;
  if (raw === "NULL") return null;
  const n = Number(raw);
  if (!Number.isNaN(n) && raw.trim() !== "") return n;
  return raw;
}

function canonicalizeTable(name: string, native: NativeTable): CanonicalTable {
  // Primary key — use primaryKeyColumns (pk-position order from adapter.primaryKey)
  // rather than filtering columns array, so composite PK ordering matches Rails.
  const pkCols = native.primaryKeyColumns;
  const primaryKey: CanonicalTable["primaryKey"] =
    pkCols.length === 0
      ? null
      : pkCols.length === 1
        ? pkCols[0]!
        : (pkCols as [string, string, ...string[]]);

  // Columns in declaration order (D1)
  const columns: CanonicalColumn[] = native.columns.map((col) => ({
    name: col.name,
    type: toCanonicalType(col.sqlType, name, col.name),
    null: col.null,
    default: coerceDefault(col.default),
    limit: col.limit,
    precision: col.precision,
    scale: col.scale,
  }));

  // Indexes: filter autoindexes (D3), sort by name (D1)
  const indexes: CanonicalIndex[] = native.indexes
    .filter((idx) => !idx.name.startsWith(AUTOINDEX_PREFIX))
    .sort((a, b) => a.name.localeCompare(b.name))
    .map((idx) => {
      if (idx.columns.length === 0) {
        throw new Error(`canonicalize: index "${idx.name}" on table "${name}" has no columns`);
      }
      return {
        name: idx.name,
        columns: idx.columns as [string, ...string[]],
        unique: idx.unique,
        where: idx.where ?? null,
      };
    });

  return { name, primaryKey, columns, indexes };
}

/**
 * Lower a NativeDump into a CanonicalSchema.
 *
 * - Filters schema_migrations / ar_internal_metadata (D2).
 * - Sorts tables by name (D1).
 * - Preserves column declaration order (D1).
 * - Filters sqlite_autoindex_* (D3).
 * - Throws on unknown SQL types (D4).
 */
export function canonicalize(native: NativeDump): CanonicalSchema {
  const tables: CanonicalTable[] = Object.keys(native)
    .filter((name) => !FILTERED_TABLES.has(name))
    .sort()
    .map((name) => canonicalizeTable(name, native[name]!));

  return { version: 1, tables };
}
